docs/ENSEMBLES.md

```diff
@@ -0,0 +1,57 @@
+# Design blueprint: ensembles
+
+We want testground to support running a test plan with instances built against
+different versions sets. For example, running the DHT test plan with 200
+instances split into groups:
+
+- Group A: 50 instances are built against go-libp2p-kad-dht/v0.3.0,
+  go-libp2p-swarm/v0.3.0
+- Group B: 50 instances are built against go-libp2p-kad-dht/v0.3.0,
+  go-libp2p-swarm/v0.4.0
+- Group C: 50 instances are built against go-libp2p-kad-dht/v0.4.0,
+  go-libp2p-swarm/v0.3.0
+- Group D: 50 instances are built against go-libp2p-kad-dht/v0.4.0,
+  go-libp2p-swarm/v0.4.0
+
+Testground has been designed to offer the primitives to compose such **test
+ensembles**. We coin the term **test ensemble** to refer to a set of test
+instances, potentially built against different dependency graphs, potentially
+running with different parameters, but orchestrated by the same runner, and
+participating under the same *run id*.
+
+Thus, all instances comprising the ensemble have visibility over one another,
+yet they are configured differently, and such variance is totally opaque.
+
+As long as the DHT plan can be compiled against all upstream versions of the
```

Comment:
Sorry if this is off topic, but this reminds me of a conversation with @aschmahmann and @daviddias during the testground unconf about how to run a test between versions that aren't fully API compatible.

For example, if I add a new public method to some type in version 0.5 that I'd like to use in a test plan, how can I test the protocol-level interactions with version 0.4 that doesn't have the new method? Since the plan depends on a method that doesn't exist in 0.4, we can't just swap out the dependencies.

I think that @daviddias was recommending that you write a shim of some kind which would stub out the new method when linked against the old version but would call the real implementation if it exists. Does anyone have thoughts on how to write that shim layer? In a language like Go it seems like the shim would have to use reflection to check for the existence of the real method. Is that the right way to do this?

FWIW, I was initially thinking I'd need to do this to test a (future) DHT with signed address records against the current version. But after thinking it over, I may just be able to add some instrumentation to the new version, so the test plan wouldn't have to call any new methods & instead could just collect the new metrics. Still, it seems like something we'll probably want to do eventually, so I figure it's worth asking.

Reply:
Reflection or build tags would be the way to go IMO. But this doesn't work always. Some behaviours cannot be shimmed -- they are novel features.

For the particular case of comparing how a DHT or pubsub performs with signed peer routing records vs. unsigned records, there shouldn't really be any public interface changes to deal with.